File: rust/src/bin/p0019.rs
// Project Euler: Problem 19

euler::run_solver!(19);

fn solve() -> String {
    compute().to_string()
}

fn compute() -> i64 {
    let common_year: Vec<i64> = vec![31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31];
    let leap_year: Vec<i64> = vec![31, 29, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31];

    // number of days in each month (from 'Jan 1901' to 'Dec 2000')
    let mut days = repeat_vec(&concat_vec(&repeat_vec(&common_year, 3), &leap_year), 25);
    // insert the number of days in the year 1900 at the beginning
    days.insert(0, 365);
    // remove 'Dec 2000'
    days.pop();

    // the `initial_state` parameter of Iterator::scan() is the day of the week on January 1, 1900
    // (0: Sunday, 1: Monday, ..., 6: Saturday)
    days.iter()
        .scan(1, |state, &x| {
            *state = (*state + x) % 7;
            Some(*state)
        })
        .filter(|&x| x == 0)
        .count() as i64
}

fn repeat_vec(v: &[i64], count: usize) -> Vec<i64> {
    v.iter().copied().cycle().take(v.len() * count).collect()
}

fn concat_vec(v1: &[i64], v2: &[i64]) -> Vec<i64> {
    [v1, v2].concat()
}

#[cfg(test)]
mod tests {
    use super::compute;

    #[test]
    fn p0019() {
        assert_eq!(compute(), 171);
    }
}

File: ocaml/src/p0019.ml
(* Project Euler: Problem 19 *)

open Core

let repeat_lst lst n =
  let rec loop i acc = if i = 0 then acc else loop (pred i) (List.append acc lst) in
  loop n []
;;

let count_sundays () =
  let common_year = [ 31; 28; 31; 30; 31; 30; 31; 31; 30; 31; 30; 31 ] in
  let leap_year = [ 31; 29; 31; 30; 31; 30; 31; 31; 30; 31; 30; 31 ] in

  (* days in month (Jan 1901 - *Nov* 2000) *)
  let days =
    List.drop_last_exn (repeat_lst (List.append (repeat_lst common_year 3) leap_year) 25)
  in

  (*
     # day of week - [0: Sunday, 1: Monday, 2: Tuesday, ..., 6: Saturday]
     # Jan 1, 1900 was Monday and assume this day is the first day. (Monday is '1 mod 7 = 1')
     # And then, the year 1900 was common year (365 days).
     # --> Jan 1, 1901 was Tuesday since (1 + 365) mod 7 = 2.
     #     Feb 1, 1901 was Firday since ((1 + 365) + 31) mod 7 = 5.
     #     ... and so on
  *)
  List.folding_map ((1 + 365) :: days) ~init:0 ~f:(fun x y -> (x + y, x + y))
  |> List.filter ~f:(fun n -> n mod 7 = 0)
  |> List.length
;;

let compute () = count_sundays ()
let solve () = compute () |> Int.to_string

(* Test *)

let%test_unit "number of Sundays" = [%test_eq: int] (compute ()) 171
